detox/android/detox/src/full/java/com/wix/detox/espresso/MultiViewActionWithResult.kt

```diff
@@ -0,0 +1,45 @@
+package com.wix.detox.espresso
+
+import android.view.View
+import androidx.test.espresso.Espresso
+import org.hamcrest.Matcher
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+
+abstract class MultiViewActionWithResult<R> : ViewActionWithResult<R> {
+    fun performOnView(viewMatcher: Matcher<View>): Any {
+        val results = ArrayList<JSONObject>()
+        var index = 0
+        var shouldCheckNextIndex = true
+
+        while (shouldCheckNextIndex) {
+            kotlin.runCatching {
+                val matcherAtIndex = DetoxMatcher.matcherForAtIndex(index, viewMatcher)
+                Espresso.onView(matcherAtIndex).perform(this)
+            }.onSuccess {
+                val actionResult = this.getResult()
+
+                lateinit var actionResultJson: JSONObject
+                try {
+                    actionResultJson = JSONObject(actionResult.toString())
+                } catch (e: JSONException) {
+                    throw e
+                }
+
+                results.add(actionResultJson)
+                index++
+            }.onFailure {
+                if (results.isEmpty()) {
+                    throw it
+                }
+
+                shouldCheckNextIndex = false
+            }
+        }
+
+        val elementsObject = JSONObject()
+        elementsObject.put("elements", JSONArray(results))
+        return if (results.size == 1) results.first() else elementsObject
```

Comment:
Better to not perform the JSON formatting etc. when it is not necessary (i.e. in the single-result case)